server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -72,9 +88,28 @@ public static DocumentMapper createEmpty(MapperService mapperService) {
             : "provided source [" + source + "] differs from mapping [" + mapping.toCompressedXContent() + "]";
     }
 
-    private void maybeLogDebug(Exception ex) {
+    private void maybeLog(Exception ex) {
         if (logger.isDebugEnabled()) {
             logger.debug("Error while parsing document: " + ex.getMessage(), ex);
+        } else {
+            final long now = Instant.now().getEpochSecond();
+            // Check without locking first, to reduce lock contention.
+            if (now - lastErrorLogEpochSecond > errorThrottlingIntervalSeconds) {
```

Comment:
Can we extract this stuff into another class? That will also allow you to provide a test implementation. Is it annoying to add a constructor parameter?

Reply:
Yeah not super excited about this.. I was contemplating whether to add a logger wrapper like `PrefixLogger` to handle throttling, @martijnvg  thoughts?

Reply:
Sounds good, moved to a separate enum, so that we can use it in more places if needed. A logger wrapper would require always creating the messages to log, even when throttled, so skipped it for now..